### PR TITLE
Add analytic _projected, _cumul2d and _fourier to Hernquist profile

### PR DESCRIPTION
**Pull request by hsinfan1996**
The expressions are given by WolframAlpha and are tested with CLMM's example notebook `examples/Paper_v1.0/validation_tests.ipynb` on the [CLMM repo](https://github.com/LSSTDESC/CLMM).

### CODE REVIEW

**Comment by hsinfan1996** [comment]
_Fourier can also be analytic, but it will only make the precision of the projected densities worse.

**Comment by damonge** [comment]
Thanks @hsinfan1996 
Not sure I understand? Why would implementing `_fourier_analytic` affect the precision of the projected profiles?

**Comment by hsinfan1996** [comment]
> Thanks @hsinfan1996 Not sure I understand? Why would implementing `_fourier_analytic` affect the precision of the projected profiles?

If _projected and _cumul2d are not available, aren't they calculated by FFT?

**Comment by damonge** [comment]
@hsinfan1996 just checking: I assume you're still working on this one, right? If not, we do need more tests (coverage decreased significantly)

**Comment by hsinfan1996** [comment]
> @hsinfan1996 just checking: I assume you're still working on this one, right? If not, we do need more tests (coverage decreased significantly)

Tests are added.

File: pyccl/halos/profiles.py [hunk 1]
```diff
@@ -919,21 +919,48 @@ class HaloProfileHernquist(HaloProfile):
...
             truncated at :math:`r = R_\\Delta` (i.e. zero at larger
             radii.
     """
     name = 'Hernquist'
 
-    def __init__(self, c_M_relation, truncated=True):
+    def __init__(self, c_M_relation,
+                 fourier_analytic=False,
+                 projected_analytic=False,
+                 cumul2d_analytic=False,
```
Comment (damonge) [inline pyccl/halos/profiles.py:940]: Can you preserve the order of `truncated`? otherwise, if someone had a code that did
```
HaloProfileHernquist(cM, False)
```
it will silently not do what they're thinking! (thus breaking the API)

File: pyccl/halos/profiles.py [hunk 2]
```diff
@@ -919,21 +919,48 @@ class HaloProfileHernquist(HaloProfile):
...
+        if cumul2d_analytic:
+            if truncated:
+                raise ValueError("Analytic cumuative 2d profile not supported "
+                                 "for truncated Hernquist. Set `truncated` or "
+                                 "`cumul2d_analytic` to `False`.")
+            self._cumul2d = self._cumul2d_analytic
         super(HaloProfileHernquist, self).__init__()
         self.update_precision_fftlog(padding_hi_fftlog=1E2,
-                                     padding_lo_fftlog=1E-2,
+                                     padding_lo_fftlog=1E-4,
```
Comment (damonge) [inline pyccl/halos/profiles.py:963]: why was this necessary?

**Review by damonge — CHANGES_REQUESTED** [review]
Thanks @hsinfan1996 . Just one comment about API breaking

File: pyccl/halos/profiles.py [hunk 3]
```diff
@@ -919,21 +919,48 @@ class HaloProfileHernquist(HaloProfile):
...
         if not isinstance(c_M_relation, Concentration):
             raise TypeError("c_M_relation must be of type `Concentration`)")
 
         self.cM = c_M_relation
         self.truncated = truncated
+        if fourier_analytic:
+            self._fourier = self._fourier_analytic
+        if projected_analytic:
+            if truncated:
+                raise ValueError("Analytic projected profile not supported "
```
Comment (damonge) [inline pyccl/halos/profiles.py:951]: One more thing, can you add a test that this error gets thrown when trying to use it with truncated?

File: pyccl/halos/profiles.py [hunk 4]
```diff
@@ -919,21 +919,48 @@ class HaloProfileHernquist(HaloProfile):
...
+                                 "for truncated Hernquist. Set `truncated` or "
+                                 "`projected_analytic` to `False`.")
+            self._projected = self._projected_analytic
+        if cumul2d_analytic:
+            if truncated:
+                raise ValueError("Analytic cumuative 2d profile not supported "
```
Comment (damonge) [inline pyccl/halos/profiles.py:957]: and this one

**Review by damonge — CHANGES_REQUESTED** [review]
And an afterthought

**Comment by hsinfan1996** [comment]
@damonge Could you wait for this PR before tagging a new release of CCL? We want to use it in CLMM but some people cannot install CCL from the source.

**Comment by damonge** [comment]
yes, I was indeed waiting for it. No worries.

File: pyccl/halos/profiles.py (hunk 2 above)
Comment (hsinfan1996) [reply pyccl/halos/profiles.py:963]: To get line 469 of `pyccl/tests/test_profiles.py` pass with a relative tolerance of 5E-3. Otherwise, the relative tolerance would need to be 0.12.

File: pyccl/halos/profiles.py (hunk 3 above)
Comment (hsinfan1996) [reply pyccl/halos/profiles.py:951]: Done.

File: pyccl/halos/profiles.py (hunk 4 above)
Comment (hsinfan1996) [reply pyccl/halos/profiles.py:957]: Done. The same test is also added for NFW.

File: pyccl/halos/profiles.py (hunk 1 above)
Comment (hsinfan1996) [reply pyccl/halos/profiles.py]: I see. Fixed.

**Review by damonge — APPROVED** [review]
LGTM!

**Comment by damonge** [comment]
(feel free to press the green button yourself @hsinfan1996 !)

**Comment by hsinfan1996** [comment]
@damonge should I choose squash and merge?

**Comment by damonge** [comment]
Always squash and merge, yes. No worries.
